modules/currency-l1/src/main/scala/org/tessellation/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,64 +52,135 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
-                    .flatMap { estimatedFee =>
-                      EstimatedFee
-                        .getUpdateHash(update, dataApplication.serializeUpdate)
-                        .map(EstimatedFeeResponse(estimatedFee, _))
-                    }
-                    .flatMap(response => Ok(response.asJson.dropNullValues))
-              }
+          validateDataUpdate(update) {
+            lastGlobalSnapshotStorage.getOrdinal.flatMap {
+              case None => InternalServerError(GL0SnapshotOrdinalUnavailable.toApplicationError)
+              case Some(ordinal) =>
+                dataApplication
+                  .estimateFee(ordinal)(update)
+                  .flatMap { estimatedFee =>
+                    EstimatedFee
+                      .getUpdateHash(update, dataApplication.serializeUpdate)
+                      .map(EstimatedFeeResponse(estimatedFee, _))
+                  }
+                  .map(_.asJson.dropNullValues)
+                  .flatMap(Ok(_))
+            }
           }
         }
         .handleUnknownError
 
     case GET -> Root / "data" =>
-      implicit val signedEncoder: Encoder[Signed[DataUpdate]] = Signed.encoder(dataApplication.dataEncoder)
-      dataUpdatesQueue.view.flatMap(Ok(_))
+      implicit val dataUpdateEncoder: Encoder[DataUpdate] = dataApplication.dataEncoder
+      dataTransactionsQueue.view.flatMap(Ok(_))
 
     case req @ POST -> Root / "data" =>
-      implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
-      req
-        .asR[Signed[DataUpdate]] {
-          _.toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _).flatMap {
-            case Left(_) => BadRequest(InvalidSignature.toApplicationError)
-            case Right(hashed) =>
-              validate(hashed.signed) {
-                dataUpdatesQueue.offer(hashed.signed) >> Ok(("hash" ->> hashed.hash.value) :: HNil)
+      def parseMultipleDataTransactions: F[DataTransactions] = {
+        implicit val dataTransactionRequest: EntityDecoder[F, DataTransactions] = dataApplication.requestDecoder
+        req.as[DataTransactions]
+      }
+
+      def parseSingleDataTransaction: F[DataTransactions] = {
+        implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
+        req
+          .as[Signed[DataUpdate]]
+          .map(NonEmptyList.of(_))
+      }
+
+      def handleDataUpdates(dataTransactions: DataTransactions, isSingleUpdate: Boolean): F[Response[F]] =
+        validateDataTransactions(dataTransactions) {
+          dataTransactions.traverse { signedTransaction =>
+            signedTransaction.value match {
+              case dataUpdate: DataUpdate =>
+                Signed(dataUpdate, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
```

Comment:
maybe you can use `.widen`? It's a bad idea to cast the whole expression. Just map/cast the required part (I believe the error?)

Reply:
I don't get it, it's failing when I try to use widen

Reply:
you need to use `widen` on Either, not on `F`